algviz/picture/recursive/rec_layout.py

```diff
@@ -0,0 +1,358 @@
+import collections
+import enum
+import math
+
+from . import elements, anchors
+from .anchors import Anchor, Coord
+
+from algviz.parser import structures
+
+class LayoutOperationError(Exception):
+    pass
+
+class AbstractLayout(object):
+
+    def __init__(self, svg_hint=None, choose_child_cls=None, ptr_targets=None):
+        self.svg_hint = svg_hint
+        self._choose_child_cls = choose_child_cls
+        self._placements = {}  # {placement: coordinate}
+        self._children = {}  # Like placements, but for sub-pictures instead of primitives
+        self._decorations = []  # these have no coordinates
+        # ptr_targets is a map from DataStructure instances to the
+        # PictureElement instances that have been placed to represent them.
+        self._ptr_targets = {} if ptr_targets is None else ptr_targets
```

Comment:
ptr_targets should be renamed as it doesn't have anything to do with pointer structures (intuitively, at least).

Reply:
Yeah, it's a mapping from DataStructure instances to the corresponding Element instances.  I named it after how I use it, not after what it is.  But I suppose that's bad.